python/core/additions/qgsfunction.py

```diff
@@ -99,78 +111,79 @@ def referencedColumns(self, node):
-                                expandargs, handlesnull)
 
-    # This doesn't really make any sense here but does when used from a decorator context
-    # so it can stay.
+    # Legacy: if args was not 'auto', parameters were passed as a list
+    params_as_list = kwargs.get("params_as_list", kwargs.get("args", "auto") != "auto")
+    f = QgsPyExpressionFunction(
+        function, name, group, helptext, usesgeometry, referenced_columns, handlesnull, params_as_list
+    )
+
     if register:
         QgsExpression.registerFunction(f)
     return f
 
 
-def qgsfunction(args='auto', group='custom', **kwargs):
+def qgsfunction(args="auto", group="custom", **kwargs):
     r"""
     Decorator function used to define a user expression function.
-
-    :param args: Number of parameters, set to 'auto' to accept a variable length of parameters.
+    :param args: DEPRECATED since QGIS 3.32. Use the "params_as_list" keyword argument instead if you want to pass parameters as a list.
     :param group: The expression group to which this expression should be added.
     :param \**kwargs:
         See below
-
     :Keyword Arguments:
         * *referenced_columns* (``list``) --
           An array of field names on which this expression works. Can be set to ``[QgsFeatureRequest.ALL_ATTRIBUTES]``. By default empty.
         * *usesgeometry* (``bool``) --
           Defines if this expression requires the geometry. By default False.
         * *handlesnull* (``bool``) --
           Defines if this expression has custom handling for NULL values. If False, the result will always be NULL as soon as any parameter is NULL. False by default.
+        * *params_as_list* (``bool``) \since QGIS 3.32 --
+          Defines if the parameters are passed to the function as a list, or if they are expanded. Default to False.
+
+    Examples:
+
+    @qgsfunction(group="custom")
+    def myfunc(values, feature, parent):
+        return values + [feature.id()]
 
-    Example:
-      @qgsfunction(2, 'test'):
-      def add(values, feature, parent):
-        pass
+    This register a function called "myfunc" in the "custom" group. It can then be called with any number of parameters
+    which will be passed as a list to the function. From the function, it is possible to access the feature and the parent
 
-    Will create and register a function in QgsExpression called 'add' in the
-    'test' group that takes two arguments.
+    >>> myfunc("a", "b", "c")
+    ["a", "b", "c", 1]
 
-    or not using feature and parent:
 
-    Example:
-      @qgsfunction(2, 'test'):
-      def add(values, *args):
-        pass
+    @qgsfunction(group="custom")
+    def myfunc2(val1, val2, context):
+        return val1 + val2
+
+    This register a function called "myfunc2" in the "custom" group. It expects exactly two parameters, val1 and val2
+    From the function, it is possible to access the feature and the parent
```

Comment:
Is it? I see `context` there
From an example point of view, it would be nice to do something with context (e.g. use an expression variable).